File: tests/python/test_soundex.py
#!/usr/bin/env python3
"""
Test suite for CThaiNLP soundex module
"""

import unittest
import os
import sys

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../..')))

from cthainlp.soundex import soundex, lk82, udom83
from cthainlp import soundex as top_soundex


class TestSoundex(unittest.TestCase):
    """Test Thai soundex algorithms"""

    def test_lk82(self):
        self.assertEqual(lk82(""), "")
        self.assertEqual(lk82("ลัก"), "ร1000")
        self.assertEqual(lk82("รัก"), "ร1000")
        self.assertEqual(lk82("รักษ์"), "ร1000")

    def test_udom83(self):
        self.assertEqual(udom83(""), "")
        self.assertEqual(udom83("ลัก"), "ร100000")
        self.assertEqual(udom83("รัก"), "ร100000")
        self.assertEqual(udom83("รักษ์"), "ร100000")

    def test_soundex_function(self):
        # Default engine is udom83
        self.assertEqual(soundex("รัก"), "ร100000")
        self.assertEqual(soundex("รัก", engine="udom83"), "ร100000")
        self.assertEqual(soundex("รัก", engine="lk82"), "ร1000")
        self.assertEqual(top_soundex("รัก"), "ร100000")

        with self.assertRaises(ValueError):
            soundex("รัก", engine="invalid_engine")


if __name__ == "__main__":
    unittest.main()

File: tests/test_soundex.c
/**
 * @file test_soundex.c
 * @brief Test suite for Thai Soundex algorithms
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "soundex.h"

int main() {
    printf("=== CThaiNLP Soundex Test Suite ===\n");
    int passed = 0;
    int total = 0;
    
    /* Test LK82 */
    {
        total++;
        char* code = soundex_lk82("ลัก");
        if (code && strcmp(code, "ร1000") == 0) {
            printf("[Test %d] LK82 'ลัก': PASS\n", total);
            passed++;
        } else {
            printf("[Test %d] LK82 'ลัก': FAIL (got %s)\n", total, code ? code : "NULL");
        }
        soundex_free(code);
    }
    
    {
        total++;
        char* code = soundex_lk82("รัก");
        if (code && strcmp(code, "ร1000") == 0) {
            printf("[Test %d] LK82 'รัก': PASS\n", total);
            passed++;
        } else {
            printf("[Test %d] LK82 'รัก': FAIL (got %s)\n", total, code ? code : "NULL");
        }
        soundex_free(code);
    }
    
    {
        total++;
        char* code = soundex_lk82("รักษ์");
        if (code && strcmp(code, "ร1000") == 0) {
            printf("[Test %d] LK82 'รักษ์': PASS\n", total);
            passed++;
        } else {
            printf("[Test %d] LK82 'รักษ์': FAIL (got %s)\n", total, code ? code : "NULL");
        }
        soundex_free(code);
    }
    
    /* Test Udom83 */
    {
        total++;
        char* code = soundex_udom83("ลัก");
        if (code && strcmp(code, "ร100000") == 0) {
            printf("[Test %d] Udom83 'ลัก': PASS\n", total);
            passed++;
        } else {
            printf("[Test %d] Udom83 'ลัก': FAIL (got %s)\n", total, code ? code : "NULL");
        }
        soundex_free(code);
    }
    
    {
        total++;
        char* code = soundex_udom83("รัก");
        if (code && strcmp(code, "ร100000") == 0) {
            printf("[Test %d] Udom83 'รัก': PASS\n", total);
            passed++;
        } else {
            printf("[Test %d] Udom83 'รัก': FAIL (got %s)\n", total, code ? code : "NULL");
        }
        soundex_free(code);
    }
    
    {
        total++;
        char* code = soundex_udom83("รักษ์");
        if (code && strcmp(code, "ร100000") == 0) {
            printf("[Test %d] Udom83 'รักษ์': PASS\n", total);
            passed++;
        } else {
            printf("[Test %d] Udom83 'รักษ์': FAIL (got %s)\n", total, code ? code : "NULL");
        }
        soundex_free(code);
    }
    
    printf("\n=== Soundex Test Summary: %d/%d passed ===\n", passed, total);
    return (passed == total) ? 0 : 1;
}
